bgpd: new vpn-policy CLI

paulzlabn opened the pull request:
PR #1739 added code to leak routes between (default VRF) VPN safi and unicast RIBs in any VRF. That set of changes included temporary CLI including `vpn-policy` blocks to specify RD/RT/label/&c. After considerable discussion, we arrived at a consensus CLI shown below.

The code of this PR implements the vpn-specific parts of this syntax:

```
router bgp <as> [vrf <FOO>]
    address-family <afi> unicast
        rd (vpn|evpn) export (AS:NN | IP:nn)
        label (vpn|evpn) export (0..1048575)
        rt (vpn|evpn) (import|export|both) RTLIST...
        nexthop vpn (import|export) (A.B.C.D | X:X::X:X)
        route-map (vpn|evpn|vrf NAME) (import|export) MAP

        [no] import|export [vpn|evpn|evpn8]
        [no] import|export vrf NAME
```

User documentation of the vpn-specific parts of the above syntax is in PR #1937

vivek-cumulus commented:
Hi @paulzlabn - the changes look fine in general from a "vpn" perspective. I have a couple of comments inline.

vivek-cumulus commented on bgpd/bgp_mplsvpn.h:
```diff
@@ -107,11 +107,8 @@ static inline int vpn_leak_to_vpn_active(struct bgp *bgp_vrf, afi_t afi,
 }
```
In this function (vpn_leak_to_vpn_active), there is no check on instance type. Of course, that is not changed in this commit, but I happened to compare with its counterpart below (vpn_leak_from_vpn_active) which has the check.

paulzlabn replied:
Added check.

vivek-cumulus commented on bgpd/bgp_mplsvpn.h:
```diff
@@ -107,11 +107,8 @@ static inline int vpn_leak_to_vpn_active(struct bgp *bgp_vrf, afi_t afi,
 }
```
Again, not modified by this commit, but isn't there a repetition in function vpn_leak_postchange() for the TOVPN case - vpn_leak_from_vrf_update_all() called twice?

paulzlabn replied:
Removed duplicate call, not sure how that sneaked in there. Thanks!

vivek-cumulus commented on bgpd/bgp_vty.c:
```diff
@@ -6369,81 +6341,40 @@ DEFPY (vpn_policy_nexthop,
-       "rt <fromvpn|tovpn|both> RTLIST...",
+DEFPY (af_rt_vpn_imexport,
+       af_rt_vpn_imexport_cmd,
+       "[no] <rt|route-target> vpn <import|export|both>$direction_str RTLIST...",
```
It appears that an add/configure operation will replace RTs and a no/delete operation will remove all RTs. Is that correct? If yes, I need to think about it because I think we'll want "additive" capability and selective delete. Of course, could be added subsequently, but IIRC, we support it now for EVPN.

paulzlabn replied:
Yes, current behavior is to replace.

Maybe "add rt vpn ... RTLIST..." or "rt vpn ... + RTLIST..."? It would probably be good to make "add" vs. "replace" semantics explicit.

Are there other parts of the FRR CLI that support additive vs. replacement behavior?

vivek-cumulus commented on bgpd/bgp_vty.c:
```diff
@@ -6477,10 +6418,18 @@ DEFUN (vpn_policy_rt,
 
 			vpn_leak_prechange(dir, afi, bgp_get_default(), bgp);
```
I don't see a check on whether import or export is enabled for this VRF or not. Without that, the pre and post change operations may walk routing tables before they bail out. Or maybe it is there some place that I missed? This comment may apply to other handlers too.

paulzlabn replied:
Added checks.

vivek-cumulus commented on bgpd/bgp_vty.c:
```diff
@@ -6477,10 +6418,18 @@ DEFUN (vpn_policy_rt,
+				bgp->vpn_policy[afi].rtlist[dir] = NULL;
+			}
 
 			vpn_leak_postchange(dir, afi, bgp_get_default(), bgp);
```
If the no/delete removes all RTs, isn't postchange operation unnecessary? It bails out if RT is not there way into the code.

paulzlabn replied:
I'm wary of removing this call here because it would not be obvious (to someone working on the code later) that this optimization can be or has been done. In the interest of avoiding future confusion by preserving a consistent design pattern of "prechange; change; postchange" I would prefer to leave this call in.

vivek-cumulus commented:
One general comment. RD_SET is currently used to denote both presence of RD and configuration of RD. When auto-RD supported is added later on, these two notions need to be separated out (or the presence checks removed).

paulzlabn commented:
@vivek-cumulus wrote:
> One general comment. RD_SET is currently used to denote both presence of RD and configuration of RD. When auto-RD supported is added later on, these two notions need to be separated out (or the presence checks removed).

Maybe we could indicate auto-RD via RD_SET + tovpn_rd.family = AF_UNSPEC.

paulzlabn commented:
NetDEF-CI wrote:
```
makeinfo --no-split -o 'frr.info' 'frr.texi'
frr.texi:5: warning: unrecognized encoding name `UTF-8'.
gmake[4]: *** [Makefile:32: frr.info] Segmentation fault (core dumped)
```

Artifact of new doc organization?

pguibert6WIND commented:
So this is the VTY for configuring RD/RT per BGP instance.
About multiple VRF for a singleBGP instance, will that ticket respond to this ?
If yes, then will vrf-policy change ?

donaldsharp commented on bgpd/bgp_vty.c:
```diff
@@ -6586,47 +6513,73 @@ DEFUN (vpn_policy_route_map,
+       bgp_imexport_vpn_cmd,
+       "[no] <import|export>$direction_str vpn",
+       "Export routes to another routing protocol\n"
+       "to VPN RIB per vpn-policy")
```
Please add the correct help strings,  we are missing some here.

paulzlabn replied:
Thanks for catching; fixed.

donaldsharp commented:
The first commit really needs a bit more explanation of what it is implementing instead of assuming the person knows what it is.

Have we updated the doc/XXX with these cli changes yet?  I do not see it yet.

paulzlabn commented:
@pguibert6WIND wrote:
 > So this is the VTY for configuring RD/RT per BGP instance.
About multiple VRF for a single BGP instance, will that ticket respond to this ?
If yes, then will vrf-policy change ?

I might misunderstand what you are asking, so if I did not answer your question, please let me know.

vrf-policy should no longer be needed.

The code in this PR supports multiple VRFs per bgpd. Here is an example from bgpd.conf:

```
router bgp 1111
   bgp router-id 1.0.0.1

   neighbor 1.0.0.2 remote-as 1111
   neighbor 1.0.0.2 update-source 1.0.0.1

   address-family ipv4 vpn
     neighbor 1.0.0.2 activate
   exit-address-family

router bgp 5001 vrf GREEN
   bgp router-id 10.0.1.1

   address-family ipv4 unicast
     rd vpn export 1:1
     rt vpn both 1:100
     import vpn
     export vpn
   exit-address-family

router bgp 5002 vrf BLUE
   bgp router-id 10.0.1.1

   address-family ipv4 unicast
     rd vpn export 2:1
     rt vpn both 1:200
     import vpn
     export vpn
   exit-address-family
```

paulzlabn commented:
@donaldsharp wrote:
> The first commit really needs a bit more explanation of what it is implementing instead of assuming the person knows what it is.
Have we updated the doc/XXX with these cli changes yet? I do not see it yet.

Updated message of first commit with more detail. Documentation of new CLI now in PR #1937
